src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/services/AuditS3Client.kt

```diff
@@ -33,19 +36,40 @@ class AuditS3Client(
       .contentMD5(md5Base64)
       .build()
 
-    s3Client.putObject(putObjectRequest, RequestBody.fromBytes(jsonBytes))
+    s3Client.putObject(putObjectRequest, RequestBody.fromBytes(parquetBytes))
   }
 
-  private fun generateFilename(auditEvent: HMPPSAuditListener.AuditEvent, auditEventJsonString: String): String {
+  private fun generateFilename(auditEvent: HMPPSAuditListener.AuditEvent): String {
     val whenDateTime = auditEvent.`when`.atZone(ZoneId.systemDefault()).toLocalDateTime()
     return "year=${whenDateTime.year}/month=${whenDateTime.monthValue}/day=${whenDateTime.dayOfMonth}/user=${auditEvent.who}/" +
-      "${getMd5HashFromJsonString(auditEventJsonString)}.json"
+      "${auditEvent.id}.parquet"
   }
 
-  private fun getMd5HashFromJsonString(jsonString: String): String {
-    val bytes = jsonString.toByteArray()
-    val md = MessageDigest.getInstance("MD5")
-    val digest = md.digest(bytes)
-    return digest.joinToString("") { "%02x".format(it) }
+  fun convertToParquetBytes(auditEvent: HMPPSAuditListener.AuditEvent, filename: String): ByteArray {
+    val schema: Schema = Schema.Parser().parse(javaClass.getResourceAsStream("/audit_event.avsc"))
+    val record: GenericRecord = GenericData.Record(schema).apply {
+      put("id", auditEvent.id?.toString())
+      put("what", auditEvent.what)
+      put("when", auditEvent.`when`.toString())
+      put("operationId", auditEvent.operationId)
+      put("subjectId", auditEvent.subjectId)
+      put("subjectType", auditEvent.subjectType)
+      put("correlationId", auditEvent.correlationId)
+      put("who", auditEvent.who)
+      put("service", auditEvent.service)
+      put("details", auditEvent.details)
+    }
+
+    val tempFilePath = Path(System.getProperty("java.io.tmpdir"), "parquet-${auditEvent.id}.parquet")
+    AvroParquetWriter.builder<GenericRecord>(tempFilePath)
```

Comment:
Do we need to delete the temp file on completion? I don't know the library but unless it says it cleans up the resources its probably a safe assumption thats something we need to ensure happens otherwise over time it will consume disk space.

Reply:
can do that

Reply:
I think we need to do something like this to ensure the temp file is cleaned up.
```
    val tempFilePath = Path(System.getProperty("java.io.tmpdir"), "parquet-${auditEvent.id}.parquet")
    return try {
      AvroParquetWriter.builder<GenericRecord>(tempFilePath)
        .withSchema(schema)
        .withCompressionCodec(CompressionCodecName.SNAPPY)
        .withConf(Configuration())
        .build().use { writer ->
          writer.write(record)
        }
      return Files.readAllBytes(tempFilePath)
    } finally {
        Files.deleteIfExists(tempFilePath)
    }
```